src/renderer/wg_engine/tvgWgPipelineRadial.cpp

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2023 the ThorVG project. All rights reserved.
+
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "tvgWgPipelineRadial.h"
+#include "tvgWgShaderSrc.h"
+
+//************************************************************************
+// WgPipelineBindGroupRadial
+//************************************************************************
+
+void WgPipelineDataRadial::updateGradient(RadialGradient* radialGradient) {
```

Comment:
So, the class logic in "WgPipelineDataRadial" is very similar to "WgPipelineDataLinear." This could lead to more maintenance costs. Can't we use a base class to share the main code and then derive from it for both Radial and Linear? It appears possible with the appropriate design for the class.

Reply:
First of all, there are not only 2 classes, but at least 3 classes with the similar logic: Solid, Radial and Linear. But that’s only from the first look they are the similar. At least they have a different uniform layout, that needs to be described separately (they have a different buffer set, sizes, indexes).
Also, we have additional pipelines with absolutely different logic such as Empty pipeline (just for stencil writing), Image pipeline. And I'm not sure, that we don’t need additional pipeline for stroking. Only what we can disaggregate, its pipeline settings and shaders description, but it will over-complicate logic. In my logic pipeline and shader is an integral, solid object.
My main idea is: **single source file is single pipeline/shader** (no meter about logic). That’s really helpful for pipeline debugging and modification. 
From other point of view, we can add additional layers of inheritance. 

In current state it looks like this:
```
 PipelineBase -> PipelineEmpty
 PipelineBase -> PipelineImage
 PipelineBase -> (PipelineSolid, PipelineLinear, PipelineRadial)
```
But it will looks like this:
```
 PipelineBase -> PipelineEmpty
 PipelineBase -> PipelineImage
 PipelineBase -> PipelineFill -> PipelineSolid
 PipelineBase -> PipelineFill -> PipelineGradient -> (PipelineLinear, PipelineRadial)
```
But it this case we will add 2 addininal layers (PipelineFill and PipelineGradient) to 3 existing classes. Is that not over complicated?

Reply:
I think the complexity wouldn't result from the number of classes/structures if it helps reduce lines of code (LOC), binary size, or improve performance, etc.

Understood. Either way, it's fine so far. Thanks.